pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -627,6 +627,286 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			return detailedErr
+		}
+		// Device mapping for global map path succeeded
+		simpleMsg, detailedMsg := volumeToMount.GenerateMsg("MapVolume.MapDevice succeeded", fmt.Sprintf("globalMapPath %q", globalMapPath))
+		verbosity := glog.Level(4)
+		og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeNormal, kevents.SuccessfulMountVolume, simpleMsg)
+		glog.V(verbosity).Infof(detailedMsg)
+
+		// Map device to pod device map path under the given pod directory using symbolic link
+		volumeMapPath, volName := blockVolumeMapper.GetPodDeviceMapPath()
+		mapErr = util.MapDevice(devicePath, volumeMapPath, volName)
+		if mapErr != nil {
+			// On failure, return error. Caller will log and retry.
+			eventErr, detailedErr := volumeToMount.GenerateError("MapVolume.MapDevice failed", mapErr)
+			og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeWarning, kevents.FailedMapVolume, eventErr.Error())
+			return detailedErr
+		}
+		// Device mapping for pod device map path succeeded
+		simpleMsg, detailedMsg = volumeToMount.GenerateMsg("MapVolume.MapDevice succeeded", fmt.Sprintf("volumeMapPath %q", volumeMapPath))
+		verbosity = glog.Level(1)
+		og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeNormal, kevents.SuccessfulMountVolume, simpleMsg)
+		glog.V(verbosity).Infof(detailedMsg)
+
+		// Update actual state of world
+		markVolMountedErr := actualStateOfWorld.MarkVolumeAsMounted(
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID,
+			volumeToMount.VolumeName,
+			nil,
+			blockVolumeMapper,
+			volumeToMount.OuterVolumeSpecName,
+			volumeToMount.VolumeGidValue)
+		if markVolMountedErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToMount.GenerateErrorDetailed("MapVolume.MarkVolumeAsMounted failed", markVolMountedErr)
+		}
+
+		return nil
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateUnmapVolumeFunc(
+	volumeToUnmount MountedVolume,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater) (func() error, string, error) {
+
+	// Get block volume unmapper plugin
+	var blockVolumeUnmapper volume.BlockVolumeUnmapper
+	blockVolumePlugin, err :=
+		og.volumePluginMgr.FindMapperPluginByName(volumeToUnmount.PluginName)
+	if err != nil {
+		return nil, "", volumeToUnmount.GenerateErrorDetailed("UnmapVolume.FindMapperPluginByName failed", err)
+	}
+	if blockVolumePlugin == nil {
+		return nil, "", volumeToUnmount.GenerateErrorDetailed("UnmapVolume.FindMapperPluginByName failed to find BlockVolumeMapper plugin. Volume plugin is nil.", nil)
+	}
+	blockVolumeUnmapper, newUnmapperErr := blockVolumePlugin.NewBlockVolumeUnmapper(
+		volumeToUnmount.InnerVolumeSpecName, volumeToUnmount.PodUID)
+	if newUnmapperErr != nil {
+		return nil, blockVolumePlugin.GetPluginName(), volumeToUnmount.GenerateErrorDetailed("UnmapVolume.NewUnmapper failed", newUnmapperErr)
+	}
+
+	return func() error {
+		// Execute tear down device
+		unmapErr := blockVolumeUnmapper.TearDownDevice()
```

Comment:
IMO this is too early to call TearDownDevice. Kubelet should make sure the device is not needed any longer (by counting symlinks) and tear it down only after that. Should it be part of GenerateUnmapDeviceFunc?

Reply:
As for the filesystem volume case, TearDown() is called at the same timing.

https://github.com/kubernetes/kubernetes/pull/51494/files#diff-450e811a4953f760ff1594ede8b2037eL521

I think TearDown() expects that the mount point is unmounted, but I saw some plugin are doing detach operation in the TearDown, because TearDown() is a only method to cleanup attached volume if the plugin is non-attachable.

But why doesn't this cause problem currently?

Yes, I think we can move this around L873 in GenerateUnmapDeviceFunc()
https://github.com/kubernetes/kubernetes/pull/51494/files#diff-450e811a4953f760ff1594ede8b2037eR873

Reply:
To be honest, TearDown is a mystery to me too. @rootfs might educate us how it's supposed to work.

Reply:
From what I've seen, the plugins that detach in the TearDown do their own refcounting, so only the last reference will detach.

But for block, the semantics for setup and teardown are different because the volume manager is the one setting up the Pod volume and global symlinks, not the plugin.

Reply:
>> IMO this is too early to call TearDownDevice. Kubelet should make sure the device is not >>needed any longer (by counting symlinks) and tear it down only after that. Should it be part of GenerateUnmapDeviceFunc?

@jsafrane @msau42 
Thank you for the explanation. Now I think the TearDownDevice() should be done after deviceOpened check is passed in GenerateUnmapDeviceFunc().
Is this correct?